handlers.go

```diff
@@ -152,9 +152,13 @@ var CreateLogFilesHandler = Handler{
 		)
 
 		if m.Cfg.FifoLogWriter != nil {
-			if err := captureFifoToFile(m.logger, logFifoPath, m.Cfg.FifoLogWriter); err != nil {
-				m.logger.Warnf("captureFifoToFile() returned %s. Continuing anyway.", err)
-			}
+			// We wrap this in a goroutine since opening the fifo queue will block
+			// until the fifo has been connected when starting firecracker
```

Comment:
I just did an experiment to see what happens when:
1. A fifo is created
2. The fifo is attempted to be opened w/ `O_RDONLY` (causing the open call to block, as expected)
3. The fifo is then unlinked before a writer opens the other side

The behavior I saw, sadly, was that the open `O_RDONLY` did not get unblocked when the fifo was unlinked.

I think this is a problem because this could actually happen with the current code if `captureFifoToFile` is blocked on the open call, something else goes wrong with VM creation causing the create to fail and the fifos to be unlinked before the VMM opens them for reading on the other side. In this case, I think we would end up having a goroutine eternally blocked waiting on the open call.

Here's the code I tested with:
```
package main

import (
        "os"
        "syscall"
        "time"
)

func main() {
        err := syscall.Mkfifo("fifo", 0700)
        if err != nil {
                panic(err)
        }

        go func() {
                time.Sleep(3*time.Second)
                err := syscall.Unlink("fifo")
                if err != nil {
                        panic(err)
                }
        }()

        _, err = os.OpenFile("fifo", os.O_RDONLY, 0600)
        if err != nil {
                panic(err)
        }
}
```

Reply:
As far as possible fixes, maybe one thing to look into is containerd's standalone `fifo` package: https://github.com/containerd/fifo

I believe it allows you to use a `context` object to control the opening of FIFOs, which could be helpful here.

Reply:
Bah, that sucks. I do have another solution but it requires a little bit more work, but should be more preferable imo

Reply:
> The behavior I saw, sadly, was that the open `O_RDONLY` did not get unblocked when the fifo was unlinked.
> 

This is not surprising. The named FIFO behaves just like any other file when it's unlinked: the directory entry is removed, and the link count decremented, but processes with open file descriptors are not impacted.

I think this problem comes down to inadequate management of filesystem resources in two different contexts: inside the jail and outside the jail. AIUI, the logic before the jailer change was correct, we just lost some state when we overwrote the configuration with the jailed paths for the FIFOs. I think we might want to provide a context-aware abstraction around file paths within the jail, allowing us to identify the appropriate path from either context.

Reply:
> This is not surprising. The named FIFO behaves just like any other file when it's unlinked: the directory entry is removed, and the link count decremented, but processes with open file descriptors are not impacted.

The part that surprises me is that AFAIK no process actually has the file descriptor open when you run that example I pasted. There is a process trying to open the fifo, but the fifo is unlinked and not opened by anything else I'm aware of. That seems bizarre but maybe it's the expected behavior.

Reply:
Yeah, POSIX specifies that open() with O_RDONLY will block forever if there isn't a writer. Once the syscall has started, there isn't really a way to short-circuit it if the file disappears. The check for existence has already succeeded and doesn't run again.

If you change your test code to add a second call to OpenFile() with os.O_WRONLY in the goroutine after the sleep, your program will run to completion.

One possibility to avoid blocking would be to add `syscall.O_NONBLOCK` to the open flags, and then retry the operation several times in case open fails with `ENXIO`.

Reply:
@xibz Are you planning to add the retries that @nmeyerhans suggested?

Reply:
@samuelkarp - Not needed as we are opening with `O_RDONLY`.

From the man page
```
A process can open a FIFO in nonblocking mode.  In this case, opening
       for read-only succeeds even if no one has opened on the write side
       yet and opening for write-only fails with ENXIO (no such device or
       address) unless the other end has already been opened.
```